File: server/src/routes/analytics.ts
import { Router, Request, Response } from "express";
import prisma from "../db";

const router = Router();

router.get("/category-breakdown", async (req: Request, res: Response) => {
  const { month, year, accountId } = req.query;

  if (!month || !year) {
    res.status(400).json({ error: "month and year are required" });
    return;
  }

  const startDate = new Date(parseInt(year as string), parseInt(month as string) - 1, 1);
  const endDate = new Date(parseInt(year as string), parseInt(month as string), 0, 23, 59, 59);

  const where: any = {
    date: { gte: startDate, lte: endDate },
    type: "debit",
  };
  if (accountId) where.accountId = accountId;

  const transactions = await prisma.transaction.findMany({
    where,
    include: { category: true },
  });

  const categoryMap: Record<string, { name: string; type: string; color: string; total: number; count: number }> = {};

  for (const t of transactions) {
    const catName = t.category?.name || "Uncategorized";
    const catType = t.category?.type || "expense";
    const catColor = t.category?.color || "#6B7280";

    if (!categoryMap[catName]) {
      categoryMap[catName] = { name: catName, type: catType, color: catColor, total: 0, count: 0 };
    }
    categoryMap[catName].total += t.amount;
    categoryMap[catName].count += 1;
  }

  res.json(Object.values(categoryMap).sort((a, b) => b.total - a.total));
});

router.get("/monthly-trend", async (req: Request, res: Response) => {
  const { months = "6" } = req.query;
  const numMonths = parseInt(months as string, 10);

  const now = new Date();
  const startDate = new Date(now.getFullYear(), now.getMonth() - numMonths + 1, 1);

  const transactions = await prisma.transaction.findMany({
    where: { date: { gte: startDate } },
    include: { category: true },
  });

  const monthlyData: Record<string, { month: string; income: number; expenses: number; investments: number }> = {};

  for (const t of transactions) {
    const d = new Date(t.date);
    const key = `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, "0")}`;

    if (!monthlyData[key]) {
      monthlyData[key] = { month: key, income: 0, expenses: 0, investments: 0 };
    }

    const catType = t.category?.type;
    if (t.type === "credit" || catType === "income") {
      monthlyData[key].income += t.amount;
    } else if (catType === "investment") {
      monthlyData[key].investments += t.amount;
    } else if (catType === "expense") {
      monthlyData[key].expenses += t.amount;
    }
  }

  const sorted = Object.values(monthlyData).sort((a, b) => a.month.localeCompare(b.month));
  res.json(sorted);
});

export default router;

File: server-py/app/routers/analytics.py
from fastapi import APIRouter, Query
from sqlmodel import Session, select

from ..database import engine
from ..models import Transaction, Category
from ..util import date_bounds, months_ago_start, parse_iso

router = APIRouter()


@router.get("/category-breakdown")
def category_breakdown(month: int | None = None, year: int | None = None,
                       accountId: str | None = None):
    gte, lt = date_bounds(month, year)
    with Session(engine) as s:
        cats = {c.id: c for c in s.exec(select(Category)).all()}
        q = select(Transaction).where(Transaction.type == "debit")
        if gte:
            q = q.where(Transaction.date >= gte, Transaction.date < lt)
        if accountId:
            q = q.where(Transaction.accountId == accountId)
        txns = s.exec(q).all()

    by_cat: dict[str, dict] = {}
    for t in txns:
        c = cats.get(t.categoryId) if t.categoryId else None
        name = c.name if c else "Uncategorized"
        ctype = c.type if c else "expense"
        color = c.color if c else "#6B7280"
        entry = by_cat.setdefault(name, {"name": name, "type": ctype, "color": color, "total": 0.0, "count": 0})
        entry["total"] += t.amount
        entry["count"] += 1
    return sorted(by_cat.values(), key=lambda x: x["total"], reverse=True)


@router.get("/monthly-trend")
def monthly_trend(months: int = Query(6)):
    start = months_ago_start(months)
    with Session(engine) as s:
        cats = {c.id: c for c in s.exec(select(Category)).all()}
        txns = s.exec(select(Transaction).where(Transaction.date >= start)).all()

    monthly: dict[str, dict] = {}
    for t in txns:
        d = parse_iso(t.date)
        key = f"{d.year:04d}-{d.month:02d}"
        m = monthly.setdefault(key, {"month": key, "income": 0.0, "expenses": 0.0, "investments": 0.0})
        ctype = cats[t.categoryId].type if t.categoryId in cats else None
        if t.type == "credit" or ctype == "income":
            m["income"] += t.amount
        elif ctype == "investment":
            m["investments"] += t.amount
        elif ctype == "expense":
            m["expenses"] += t.amount
    return sorted(monthly.values(), key=lambda x: x["month"])
